torch/csrc/dynamo/guards.cpp

```diff
@@ -922,6 +922,26 @@ class GLOBAL_STATE : public LeafGuard {
   // TODO(jansel): we should guard on more state as inductor starts using it
 };
 
+class DATA_PTR_MATCH : public LeafGuard {
+ public:
+  DATA_PTR_MATCH(py::object data_ptr, py::object verbose_code_parts)
+      : LeafGuard(verbose_code_parts), _data_ptr(data_ptr) {}
+
+  bool check_nopybind(PyObject* value) override { // borrowed ref
+    PyObject* data_ptr_method =
+        PyObject_GetAttrString(value, "data_ptr"); // new ref
+    PyObject* data_ptr = PyObject_CallNoArgs(data_ptr_method); // new ref
+    bool result = PyObject_RichCompareBool(data_ptr, _data_ptr.ptr(), Py_EQ);
+    Py_DECREF(data_ptr);
+    Py_DECREF(data_ptr_method);
```

Comment:
If value is a Tensor, we should be able to *much* faster than this by using the C++ apis.

Reply:
Done. Ready for re-review.